Ruby/Ruby.sublime-syntax

```diff
@@ -203,8 +203,70 @@ contexts:
       scope: constant.language.ruby
     - match: '\b(__(FILE|LINE|ENCODING)__|self)\b(?![?!])'
       scope: variable.language.ruby
-    - match: '\b(0[xX]\h(_?\h)*|\d(_?\d)*(\.(?![^[:space:][:digit:]])(_?\d)*)?([eE][-+]?\d(_?\d)*)?|0[bB][01]+)\b'
-      scope: constant.numeric.ruby
+    # numbers with 'ir' suffix are invalid
+    - match: '\b(?:\d(?:_?\d)*(?:\.\d(?:_?\d)*)?(?:[eE][-+]?\d(?:_?\d)*)?|0[dDxXoObB]?\h(?:_?\h)*)ir\b'
+      scope: invalid.illegal.ruby
+    - match: '\b(0[dD])\d(?:_?\d)*(r)?(i)?\b'
+      scope: constant.numeric.integer.decimal.ruby
+      captures:
+        1: punctuation.definition.numeric.decimal.ruby
+        2: storage.type.numeric.rational.ruby
+        3: storage.type.numeric.imaginary.ruby
+    - match: '\b(0[xX])\h(?:_?\h)*(r)?(i)?\b'
+      scope: constant.numeric.integer.hexadecimal.ruby
+      captures:
+        1: punctuation.definition.numeric.hexadecimal.ruby
+        2: storage.type.numeric.rational.ruby
+        3: storage.type.numeric.imaginary.ruby
+    - match: '\b(0[oO]?)[0-7](?:_?[0-7])*(r)?(i)?\b'
+      scope: constant.numeric.integer.octal.ruby
+      captures:
+        1: punctuation.definition.numeric.octal.ruby
+        2: storage.type.numeric.rational.ruby
+        3: storage.type.numeric.imaginary.ruby
+    - match: '\b(0[bB])[01](?:_?[01])*(r)?(i)?\b'
+      scope: constant.numeric.integer.binary.ruby
+      captures:
+        1: punctuation.definition.numeric.binary.ruby
+        2: storage.type.numeric.rational.ruby
+        3: storage.type.numeric.imaginary.ruby
+    # 1.1e1i, 1e1i
+    - match: '\b\d(?:_?\d)*(?:(\.)\d(?:_?\d)*)?[eE][-+]?\d(?:_?\d)*(i)\b'
+      scope: constant.numeric.complex.imaginary.ruby
+      captures:
+        1: punctuation.separator.decimal.ruby
+        2: storage.type.numeric.imaginary.ruby
+    # 1.1ri, 1.1i, 1ri, 1i
+    - match: '\b\d(?:_?\d)*(?:(\.)\d(?:_?\d)*)?(r)?(i)\b'
+      scope: constant.numeric.complex.imaginary.ruby
+      captures:
+        1: punctuation.separator.decimal.ruby
+        2: storage.type.numeric.rational.ruby
+        3: storage.type.numeric.imaginary.ruby
+    # 1.i
+    - match: '\b\d(?:_?\d)*(\.)(i)\b'
+      scope: constant.numeric.complex.imaginary.ruby
+      captures:
+        1: punctuation.separator.decimal.ruby
+        2: storage.type.numeric.imaginary.ruby
```

Comment:
What do you think about merging the two rules for `# 1.1ri, 1.1i, 1ri, 1i` and `# 1.i` to:

```yaml
    # 1.1ri, 1.1i, 1.i, 1ri, 1i
    - match: '\b\d(?:_?\d)*(?:(\.)(?:\d(?:_?\d)*)?(r)?)?(i)\b'
      scope: constant.numeric.complex.imaginary.ruby
      captures:
        1: punctuation.separator.decimal.ruby
        2: storage.type.numeric.rational.ruby
        3: storage.type.numeric.imaginary.ruby
```

Should add some test cases for the `1.i` and `1.0i`, then:

```
1.i         #=> (0+1i)
#^^^ constant.numeric.complex.imaginary
#  ^ storage.type.numeric.imaginary
 1.0i        #=> (0+1i)
#^^^^ constant.numeric.complex.imaginary
#   ^ storage.type.numeric.imaginary
```

Reply:
The problem with that regex is that it would also match `1.ri` (which is illegal), while for this my current implementation only highlights the first part of it, but not the `ri`, so that users at least get a hint that this is actually a syntax error. If you prefer to not take that into account and favor the combined rule, I could change it though.
I will add the additional test cases.

Reply:
Didn't realize that. We should of course not break highlighting due to some merges. I just came around with that with Perl's numbers in mind, which I also adapted to Python in #2136. I was able to tweak performance by about 14% by thouroghly merging regexp patterns.

Note:

I didn't use variables in Perl, but found them quite useful in manner of readability, while working on #2136.

Based on those, we could create one single match rule for all imaginary numbers including some invalid handling


```yml
variables:
  digit: (?:\d(?:_?\d)*)
  exponent: (?:[Ee][-+]?{{digit}})

  # ...

contexts:

  # ...

    # complex imaginary numbers
    - match: |-
        \b(?x:
            {{digit}} (?:
                # 1ri, 1.1ri, 1i, 1.1i | 1e1i, 1.1e1i
              (?: (\.) {{digit}} )? (?: (r) | {{exponent}}? )
              | # 1.ri, 1.1ri, 1.e1ri, 1.1e1ri | 1.i 1.e1i
              (\.) (?: {{digit}}? {{exponent}}? (r) | ({{exponent}})? )
            )
        )(i)\b
      scope: constant.numeric.complex.imaginary.ruby
      captures:
        1: punctuation.separator.decimal.ruby
        2: storage.type.numeric.rational.ruby
        3: punctuation.separator.decimal.ruby
        4: invalid.illegal.numeric.rational.ruby
        5: invalid.illegal.numeric.exponent.ruby
        6: storage.type.numeric.imaginary.ruby
```

_Note: I am about to prepare a complete proposal with some test cases for you._